pytorch_lightning/utilities/fetching.py

```diff
@@ -173,9 +169,7 @@ def _attach_data_fetcher_fn(loader: DataLoader):
 
         apply_to_collection(self.loaders, (DataLoader, CycleIterator), _attach_data_fetcher_fn)
 
-    def __iter__(self) -> Generator[Tuple[Any, bool], None, None]:
-        if self.dataloader is None:
-            raise MisconfigurationException("The iterate hasn't been provided. HINT: Did you call setup function ?.")
+    def __iter__(self) -> "AbstractDataFetcher":
```

Comment:
What about annotating with `Self` here and only falling back to this for older python versions?

Reply:
I tried this trick:

```python
try:
    from typing_extensions import Self
except ImportError:
    # workaround for Python 3.7.
    # see https://www.python.org/dev/peps/pep-0673/
    from typing import TypeVar

    Self = TypeVar("AbstractDataFetcher", bound="AbstractDataFetcher")
```

But getting:

```python
pytorch_lightning/utilities/fetching.py:43: error: String argument 1 "AbstractDataFetcher" to TypeVar(...) does not match variable name "Self"  [misc]
pytorch_lightning/utilities/fetching.py:43: error: Incompatible types in assignment (expression has type "TypeVar", variable has type "_SpecialForm")  [assignment]
pytorch_lightning/utilities/fetching.py:182: error: Variable "typing_extensions.Self" is not valid as a type  [valid-type]
pytorch_lightning/utilities/fetching.py:182: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases

```

Reply:
I see. Unfortunately this is a very cryptic error message. So I also have no idea how to resolve it